clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/servergroup/ops/CreateAzureServerGroupAtomicOperation.groovy

```diff
@@ -68,45 +69,80 @@ class CreateAzureServerGroupAtomicOperation implements AtomicOperation<Map> {
+      }
+
+      virtualNetworkName = appGatewayDescription.vnet
+      if (description.vnet && description.vnet != virtualNetworkName) {
+        throw new RuntimeException("Invalid load balancer was selected; $description.appGatewayName does not exist")
+      }
 
-      // Compute the next subnet address prefix using the cached vnet and a random generated seed
       def vnetDescription = networkProvider.get(description.accountName, description.region, virtualNetworkName)
-      Random rand = new Random()
-      def nextSubnetAddressPrefix = AzureVirtualNetworkDescription.getNextSubnetAddressPrefix(vnetDescription, rand.nextInt(vnetDescription?.maxSubnets ?: 1))
-      subnetName = AzureUtilities.getSubnetName(virtualNetworkName, nextSubnetAddressPrefix)
-
-      // we'll do a final check to make sure that the subnet can be created before we pass it in the deployment template
-      def vnet = description.credentials.networkClient.getVirtualNetwork(resourceGroupName, virtualNetworkName)
-
-      if (!subnetName || vnet?.subnets?.find {it.name == subnetName}) {
-        // virtualNetworkName is not yet in the cache or the subnet we try to create already exists; we'll use the current vnet
-        //   we just got to re-compute the next subnet
-        vnetDescription = AzureVirtualNetworkDescription.getDescriptionForVirtualNetwork(vnet)
-        nextSubnetAddressPrefix = AzureVirtualNetworkDescription.getNextSubnetAddressPrefix(vnetDescription, rand.nextInt(vnetDescription?.maxSubnets ?: 1))
-        subnetName = AzureUtilities.getSubnetName(virtualNetworkName, nextSubnetAddressPrefix)
+
+      if (!vnetDescription) {
+        throw new RuntimeException("Selected virtual network $virtualNetworkName does not exist")
       }
 
-      String subnetId = description.credentials.networkClient.createSubnet(resourceGroupName,
-        virtualNetworkName,
-        subnetName,
-        nextSubnetAddressPrefix,
-        description.securityGroup?.name)
+      if (description.subnet) {
+        task.updateStatus(BASE_PHASE, "Using virtual network $virtualNetworkName and subnet $description.subnet for server group $description.name")
+
+        // we will try to associate the server group with the selected virtual network and subnet
+        description.hasNewSubnet = false
+
+        // subnet is valid only if it exists within the selected vnet and it's unassigned or all the associations are NOT application gateways
+        subnetId = vnetDescription.subnets?.find { subnet ->
+          (subnet.name == description.subnet) && (!subnet.connectedDevices || !subnet.connectedDevices.find {it.type == "applicationGateways"})
+        }?.resourceId
+
+        if (!subnetId) {
+          throw new RuntimeException("Selected subnet $description.subnet in virtual network $description.vnet is not valid")
+        }
+      } else {
+        task.updateStatus(BASE_PHASE, "Creating subnet for server group")
```

Comment:
same thing here as with the UpsertAzureAppGateway. We shouldn't create a subnet if the customer does not want us doing that.